metals/src/main/scala/scala/meta/internal/metals/BuildTargets.scala

```diff
@@ -515,68 +393,45 @@ final class BuildTargets(
   private def computeInverseDependencies(
       target: BuildTargetIdentifier
   ): BuildTargets.InverseDependencies = {
-    BuildTargets.inverseDependencies(List(target), inverseDependencies.get)
-  }
-
-  def addDependencySource(
-      sourcesJar: AbsolutePath,
-      target: BuildTargetIdentifier
-  ): Unit = {
-    sourceJarNameToJarFile(sourcesJar.filename) = sourcesJar
-    val acc = inverseDependencySources.getOrElse(sourcesJar, Set.empty)
-    inverseDependencySources(sourcesJar) = acc + target
+    BuildTargets.inverseDependencies(
+      List(target),
+      id => data.fromOptions(_.inverseDependencies.get(id))
+    )
   }
 
-  def sourceJarFile(sourceJarName: String): Option[AbsolutePath] = {
-    sourceJarNameToJarFile.get(sourceJarName)
-  }
+  def sourceJarFile(sourceJarName: String): Option[AbsolutePath] =
+    data.fromOptions(_.sourceJarNameToJarFile.get(sourceJarName))
 
   def inverseDependencySource(
       sourceJar: AbsolutePath
   ): collection.Set[BuildTargetIdentifier] = {
-    inverseDependencySources.get(sourceJar).getOrElse(Set.empty)
+    data
+      .fromOptions(_.inverseDependencySources.get(sourceJar))
+      .getOrElse(Set.empty)
   }
 
-  def addSourceRoot(root: AbsolutePath): Unit = {
-    isSourceRoot.add(root)
-  }
   def sourceRoots: Iterable[AbsolutePath] = {
-    isSourceRoot.asScala
+    data.iterable.flatMap(_.isSourceRoot.asScala)
   }
 
   def isInsideSourceRoot(path: AbsolutePath): Boolean = {
-    !isSourceRoot.contains(path) &&
-    isSourceRoot.asScala.exists { root => path.toNIO.startsWith(root.toNIO) }
-  }
-
-  def resetConnections(
-      idToConn: List[(BuildTargetIdentifier, BuildServerConnection)]
-  ): Unit = {
-    targetToConnection.clear()
-    idToConn.foreach { case (id, conn) => targetToConnection.put(id, conn) }
+    data.iterator.exists(_.isSourceRoot.contains(path)) &&
+    data.fromIterators(_.isSourceRoot.asScala.iterator).exists { root =>
+      path.toNIO.startsWith(root.toNIO)
+    }
   }
 
   def buildServerOf(
       id: BuildTargetIdentifier
-  ): Option[BuildServerConnection] = {
-    ammoniteBuildServer(id).orElse(targetToConnection.get(id))
-  }
+  ): Option[BuildServerConnection] =
+    data.fromOptions(_.targetToConnection.get(id))
+
+  def addData(data: TargetData): Unit =
+    this.data = BuildTargets.DataSeq(data :: this.data.list)
```

Comment:
This is more like `setData`, should it be synchronized?

Reply:
It's only called when `MetalsLanguageServer` is instantiated (or during the initialize request), so there's no concurrency issue for now.

But I just synchronized it anyway, just in case this is called in a different way in the future.